api/v1alpha1/raycluster_webhook.go

```diff
@@ -0,0 +1,310 @@
+// SetupWebhookWithManager creates and registers this webhook with the manager.
+func (r *RayCluster) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+//+kubebuilder:webhook:path=/mutate-distributed-compute-dominodatalab-com-v1alpha1-raycluster,mutating=true,failurePolicy=fail,sideEffects=None,groups=distributed-compute.dominodatalab.com,resources=rayclusters,verbs=create;update,versions=v1alpha1,name=mraycluster.kb.io,admissionReviewVersions={v1,v1beta1}
+
+var _ webhook.Defaulter = &RayCluster{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *RayCluster) Default() {
+	log := logger.WithValues("raycluster", client.ObjectKeyFromObject(r))
+	log.Info("applying defaults")
+
+	if r.Spec.Port == 0 {
+		log.Info("setting default port", "value", defaultPort)
+		r.Spec.Port = defaultPort
+	}
+	if r.Spec.RedisShardPorts == nil {
+		log.Info("setting default redis shard ports", "value", defaultRedisShardPorts)
+		r.Spec.RedisShardPorts = defaultRedisShardPorts
+	}
+	if r.Spec.ClientServerPort == 0 {
+		log.Info("setting default client server port", "value", defaultClientServerPort)
+		r.Spec.ClientServerPort = defaultClientServerPort
+	}
+	if r.Spec.ObjectManagerPort == 0 {
+		log.Info("setting default object manager port", "value", defaultObjectManagerPort)
+		r.Spec.ObjectManagerPort = defaultObjectManagerPort
+	}
+	if r.Spec.NodeManagerPort == 0 {
+		log.Info("setting default node manager port", "value", defaultNodeManagerPort)
+		r.Spec.NodeManagerPort = defaultNodeManagerPort
+	}
+	if r.Spec.DashboardPort == 0 {
+		log.Info("setting default dashboard port", "value", defaultDashboardPort)
+		r.Spec.DashboardPort = defaultDashboardPort
+	}
+	if r.Spec.EnableDashboard == nil {
+		log.Info("setting enable dashboard flag", "value", *defaultEnableDashboard)
+		r.Spec.EnableDashboard = defaultEnableDashboard
+	}
+	if r.Spec.EnableNetworkPolicy == nil {
+		log.Info("setting enable network policy flag", "value", *defaultEnableNetworkPolicy)
+		r.Spec.EnableNetworkPolicy = defaultEnableNetworkPolicy
+	}
+	if r.Spec.NetworkPolicyClientLabels == nil {
+		log.Info("setting default network policy client labels", "value", defaultNetworkPolicyClientLabels)
+		r.Spec.NetworkPolicyClientLabels = defaultNetworkPolicyClientLabels
+	}
+	if r.Spec.Worker.Replicas == nil {
+		log.Info("setting default worker replicas", "value", *defaultWorkerReplicas)
+		r.Spec.Worker.Replicas = defaultWorkerReplicas
+	}
+
+	if r.Spec.Image == nil {
+		log.Info("setting default image", "value", *defaultImage)
+		r.Spec.Image = defaultImage
+	} else {
+		if r.Spec.Image.Repository == "" {
+			log.Info("setting default image repository", "value", defaultImage.Repository)
+			r.Spec.Image.Repository = defaultImage.Repository
+		}
+		if r.Spec.Image.Tag == "" {
+			log.Info("setting default image tag", "value", defaultImage.Tag)
+			r.Spec.Image.Tag = defaultImage.Tag
+		}
```

Comment:
This seems a bit odd, if I define the Image and want the repo or tag empty shouldn't it leave it be?

Reply:
### Repo

```go
type OCIImageDefinition struct {
	// Registry where the container image is hosted.
	Registry string `json:"registry,omitempty"`

	// Repository where the container image is stored.
	Repository string `json:"repository,omitempty"`

	// Tag points to a specific container image variant.
	Tag string `json:"tag,omitempty"`

	// PullPolicy used to fetch container image.
	PullPolicy corev1.PullPolicy `json:"pullPolicy,omitempty"`
}
```

To my knowledge, it's fine if a `Registry` is empty, but a`Repository` cannot be. Are you thinking of the former?

### Tag

I didn't think of it before but I guess it's valid for `Tag` to be blank. The `latest` tag will be applied to the image reference when it's missing:

https://github.com/dominodatalab/distributed-compute-operator/blob/1b2286d9f24e0c715b8941b04b4d21c954dba9e6/pkg/util/util.go#L44

I can remove this default an allow an empty string to propagate but it's worth noting this will create an inconsistency between the image reference and what's in the spec, right?

```yaml
apiVersion: distributed-compute.dominodatalab.com/v1alpha1
kind: RayCluster
metadata:
  name: steve
spec:
  image:
    repository: steve-o
    # no indication of "latest"

apiVersion: apps/v1
kind: Deployment
metadata:
  name: steve-o-ray
spec:
  template:
    spec:
      containers:
      - image: steve-o:latest
```

Does that implicit bother you?

Reply:
Yep, I was thinking of `registry` for the first one.

> I can remove this default an allow an empty string to propagate but it's worth noting this will create an inconsistency between the image reference and what's in the spec, right?

Yes, though if it implicitly changes it to `v1.2.0-cpu` that would strike me a bit odd as well. Maybe we just hard-reject empty tags given it's a pretty bad idea overall.

Reply:
modified accordingly